app/services/course/duplication_service.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+# Provides a service object for duplicating courses
+class Course::DuplicationService
+  # Constructor for the duplication service object.
+  #
+  # @param [Course] current_course The current course.
+  # @param [Hash] duplication_params A hash of duplication parameters.
+  # @param [Array] all_objects All the objects in the course.
+  # @param [Array] selected_objects The objects to duplicate.
+  def initialize(current_course, duplication_params = {}, all_objects = [], selected_objects = [])
+    @current_course = current_course
+    @all_objects = all_objects.append(current_course)
+    @selected_objects = selected_objects.append(current_course)
+    @duplication_params = duplication_params
+    @duplication_params[:new_course_start_date] = Time.zone.parse(
+      duplication_params[:new_course_start_date]
+    )
+  end
+
+  # Duplicate the course with the duplicator.
+  # Do not just pass in @selected_objects or object parents could be set incorrectly.
+  #
+  # @return [Boolean] Whether the duplication succeeded.
+  def duplicate
+    @new_course = duplicator.duplicate(@current_course)
+    @new_course.save!
```

Comment:
Use `save` instead of `save!`, `save!` will raise an exception and the return type is not boolean then.